File: go/generations_test.go
package buble

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"testing"
	"time"
)

func TestCreatesFlatGenerationBody(t *testing.T) {
	var body map[string]any
	client, server := newTestClient(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/generations" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatalf("decode request: %v", err)
		}
		writeJSON(t, w, http.StatusCreated, map[string]any{"data": map[string]any{"id": "task_1", "status": "pending"}})
	})
	defer server.Close()

	result, err := client.Generations.Create(context.Background(), &CreateGenerationRequest{
		Model:  "google/nano-banana",
		Mode:   "text_to_image",
		Prompt: "A test image",
		Params: map[string]any{
			"aspect_ratio":  "1:1",
			"output_format": "png",
		},
	})
	if err != nil {
		t.Fatalf("Create returned error: %v", err)
	}
	if result.Data.ID != "task_1" {
		t.Fatalf("unexpected task: %#v", result.Data)
	}
	if body["model"] != "google/nano-banana" || body["mode"] != "text_to_image" || body["aspect_ratio"] != "1:1" {
		t.Fatalf("request body was not flat public JSON: %#v", body)
	}
	if _, ok := body["params"]; ok {
		t.Fatalf("request body leaked Params field: %#v", body)
	}
}

func TestWaitUntilSuccess(t *testing.T) {
	statuses := []TaskStatus{StatusPending, StatusProcessing, StatusSuccess}
	client, server := newTestClient(func(w http.ResponseWriter, r *http.Request) {
		status := statuses[0]
		statuses = statuses[1:]
		writeJSON(t, w, http.StatusOK, map[string]any{"data": map[string]any{"id": "task_1", "status": status}})
	})
	defer server.Close()

	result, err := client.Generations.Wait(
		context.Background(),
		"task_1",
		WithWaitInterval(time.Millisecond),
		WithWaitTimeout(time.Second),
	)
	if err != nil {
		t.Fatalf("Wait returned error: %v", err)
	}
	if result.Data.Status != StatusSuccess {
		t.Fatalf("unexpected status: %s", result.Data.Status)
	}
}

func TestWaitFailedRaisesGenerationFailedError(t *testing.T) {
	client, server := newTestClient(func(w http.ResponseWriter, r *http.Request) {
		writeJSON(t, w, http.StatusOK, map[string]any{
			"data": map[string]any{
				"id":     "task_1",
				"status": "failed",
				"error":  map[string]any{"message": "Generation failed."},
			},
		})
	})
	defer server.Close()

	_, err := client.Generations.Wait(context.Background(), "task_1", WithWaitTimeout(time.Second))
	genErr := &GenerationFailedError{}
	if !errors.As(err, &genErr) {
		t.Fatalf("expected GenerationFailedError, got %T: %v", err, err)
	}
}

File: ruby/test/generations_test.rb
# frozen_string_literal: true

require_relative 'test_helper'

class GenerationsTest < Minitest::Test
  def test_creates_flat_generation_body
    transport = FakeTransport.new
    transport.enqueue_response('data' => { 'id' => 'task_1' })
    resource = Buble::GenerationsResource.new(transport)

    resource.create(
      model: 'google/nano-banana',
      mode: 'text_to_image',
      prompt: 'hello',
      aspect_ratio: '1:1',
      output_format: 'png'
    )

    assert_equal '/api/v1/generations', transport.requests[0][:path]
    assert_equal(
      {
        'model' => 'google/nano-banana',
        'mode' => 'text_to_image',
        'prompt' => 'hello',
        'aspect_ratio' => '1:1',
        'output_format' => 'png'
      },
      transport.requests[0][:body]
    )
  end

  def test_rejects_internal_generation_fields
    resource = Buble::GenerationsResource.new(FakeTransport.new)

    error = assert_raises(Buble::UnsupportedGenerationFieldError) do
      resource.create(model: 'model', media_type: 'image')
    end

    assert_equal 'media_type', error.field
  end

  def test_waits_until_success
    transport = FakeTransport.new
    transport.enqueue_response('data' => { 'id' => 'task_1', 'status' => 'processing' })
    transport.enqueue_response('data' => { 'id' => 'task_1', 'status' => 'success' })
    resource = Buble::GenerationsResource.new(transport)

    result = resource.wait('task_1', interval: 0, timeout: 1)

    assert_equal 'success', result.dig('data', 'status')
    assert_equal '/api/v1/generations/task_1', transport.requests[0][:path]
  end

  def test_raises_on_failed_generation
    transport = FakeTransport.new
    transport.enqueue_response('data' => {
                                 'id' => 'task_1',
                                 'status' => 'failed',
                                 'error' => { 'message' => 'provider failed' }
                               })
    resource = Buble::GenerationsResource.new(transport)

    error = assert_raises(Buble::GenerationFailedError) do
      resource.wait('task_1', interval: 0, timeout: 1)
    end

    assert_equal 'provider failed', error.message
  end
end
